### PR TITLE
feat: universal-router-polish

### PR DESCRIPTION
**Pull request by ChefBingbong**
### Summary
🚀🔐🛠️

This pull request adds support for cross-chain swapping and gasless approvals using the universal router and the permit2 contracts in the web app. It also refactors and simplifies some components and hooks, updates some dependencies and settings, and adds some translations and tests.

> _To swap tokens across different chains_
> _We added some SDKs and some brains_
> _We used `permit2`_
> _To save gas and time too_
> _And we fixed some bugs and some pains_

### Walkthrough
*  Add and use the permit2-sdk package to enable gasless approval of tokens using the permit2 contract ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R56), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-9aa3afd93d70a5b36f9456b93a5d00824f4197ee6bacb4288f8e11ab039fc7f9R1-R143), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-db20320ab1c38d09edd199a07dcec64dc27aa7f526b5d43216d6a174e4772ab1R1-R117), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-64cce548f87cb10257d780d98563d8c44d26c763287384a63492bc1dc27e56ceL3-R21), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL1-R8), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L3-R12), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R52-R61), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L95-R111), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R132), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R209-R210), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L251-R256), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L30-R30), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L49-R66), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL38-R38), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL49-R49), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-6257ad63d7362ad03e9d623477858447ee80997c2cbf8beb8e2d8396cbff1f11L3-R16), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-eb2ab983e2cefc22516cb7814c86346f4b2bd2b971980952868173095d48f459L2715-R2717), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-793ae5af075d7cec4152c6f9ef87e4091736ec8a1344f1e06daccc4935f52e6eL1-L-1), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-ec7a712e0ce7ea429f949f8b0b80f5b59a47d4cf48a59622bf2c4edd9508049dR1-R39))
*  Add and use the universal-router-sdk package to enable cross-chain swapping of tokens using the universal router contract ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R50), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-64cce548f87cb10257d780d98563d8c44d26c763287384a63492bc1dc27e56ceL3-R21), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL1-R8), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L3-R12), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L76-R95), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L3-R10), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L49-R66))
*  Modify the ConfirmSwapModal component to handle the approval and swap flow based on the permit allowance, the approval state, and the swap callback ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL45-R213), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL67-R223), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR237-R246), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL88-R254), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL95-R276), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL163-R334), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR393-R395), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL228-R402), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL246-R419))
*  Modify the SwapCommitButton component to use the permit2 contract and the universal router contract for approving and swapping tokens ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L54-R69), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L76-R95), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L95-R111), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R132), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R209-R210), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L251-R256), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L349-R349))
*  Modify the useSwapCallback hook to use the permit2 contract and the universal router contract for generating and sending the swap transaction ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL17-R18), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL38-R38), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL49-R49), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL66-R65), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL78-R77))
*  Modify the useSwapCallArguments hook to use the PancakeUniSwapRouter.swapERC20CallParameters function for generating the call data for the swap transaction ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L3-R10), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L30-R30), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-77157a86e87d5ceb7fb217f4d7adcedb558954f2133bd0e502299641aba9b7c5L49-R66))
*  Add the ApprovalSteps component to render the steps for the approval and swap flow ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c227d318fc52f432f18494faf11d07f3e01dfdb179e4129e7191048f9b51dda8R33-R50), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c227d318fc52f432f18494faf11d07f3e01dfdb179e4129e7191048f9b51dda8L58-R75))
*  Modify the useWallchainApi hook and the useWallchainSwapCallArguments hook to pass undefined instead of deadline as the third argument, as the deadline is passed as the fifth argument instead ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-bfe39efb83c5d5d051644d7ebcf06f35473f3a7d27706eae1706fd5e95acac8cL132-R132), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL78-R77))
*  Modify the dev script in the package.json file to increase the concurrency limit from 21 to 23 ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L14-R14))
*  Modify the dev script in the web app package.json to use port 3001 instead of the default port 3000 ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L7-R7))
*  Add a new setting to the VSCode configuration file to ignore the spelling of the word "bips" ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357L4-R5))
*  Remove unused imports and props from various components and hooks ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-f4ff9e7ae71a59353f06ecb29ce0023db6c69870cd4cea6fa12d4301ee52dcf1L17), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-64cce548f87cb10257d780d98563d8c44d26c763287384a63492bc1dc27e56ceL154-L166), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-64cce548f87cb10257d780d98563d8c44d26c763287384a63492bc1dc27e56ceL174-L176), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-64cce548f87cb10257d780d98563d8c44d26c763287384a63492bc1dc27e56ceL200-R189), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL31-R34), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dL204), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L207-L219), [link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L229-R236))
*  Comment out the gas calculation for the callWithGasPrice hook, as it was causing very high gas estimates ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-f4ff9e7ae71a59353f06ecb29ce0023db6c69870cd4cea6fa12d4301ee52dcf1L97-R97))
*  Comment out the resetSwapFlow and startSwapFlow calls from the MMCommitButton component, as they are no longer needed ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-64cce548f87cb10257d780d98563d8c44d26c763287384a63492bc1dc27e56ceL200-R189))
*  Add a console log statement to the useSendSwapTransaction hook to debug the gas estimate for the swap transaction ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739R94))
*  Add two new translations to the translations.json file for the messages "Reset approval on USDT." and "Permit %symbol%" ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-eb2ab983e2cefc22516cb7814c86346f4b2bd2b971980952868173095d48f459L2715-R2717))
*  Add a new file to the permit2-sdk package: .gitattributes, which specifies that the .sol files should be recognized as Solidity language by GitHub ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-793ae5af075d7cec4152c6f9ef87e4091736ec8a1344f1e06daccc4935f52e6eL1-L-1))
*  Add a new file to the permit2-sdk package: .github/workflows/forge-tests.yml, which defines a GitHub action workflow to run unit tests using the Foundry toolchain ([link](https://github.com/pancakeswap/pancake-frontend/pull/7968/files?diff=unified&w=0#diff-ec7a712e0ce7ea429f949f8b0b80f5b59a47d4cf48a59622bf2c4edd9508049dR1-R39))

### CODE REVIEW

**Comment by ChefBingbong** [comment]
> Looks like most commits that recently pushed to develop reverted with this pr. Is this intended?

nope thanks for raising meant to base with jacksons branch. changing now

File: packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts [hunk 1]
```diff
@@ -184,23 +174,29 @@ async function addV3Swap(
...
 
     // similar to encodeV3Swap only we dont need to add a case for signle hop. by using ecodeMixedRoutePath
     // we can get the parthStr for all cases
     if (trade.tradeType === TradeType.EXACT_INPUT) {
-      const exactInputSingleParams = [recipient, amountIn, amountOut, path, payerIsUser]
+      const exactInputSingleParams = [
+        recipient,
+        amountIn,
+        performAggregatedSlippageCheck ? 0n : amountOut,
+        ,
```
Comment (ChefMomota) [inline packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:196]: ?

Comment (ChefBingbong) [reply packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts]: this is samme method as in smart router sdk. this is correct. im pretty sure this aggregated slippage check is to help make sure trade doesnt fail when there is more than one swap. amountOutMin is pretty much similar to slippage so '0n' here is just a fallback i reckon to allow slippage for full range if multiple trades tto prevent execution from failing.

Comment (ChefBingbong) [reply packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts]: is only done on trades with 2 swaps,

**Pull request merged**